lib/src/puppetdb-cli.cc

```diff
@@ -107,7 +112,83 @@ PuppetDBConn::parseServerUrls(const json::JsonContainer& config) {
     }
 }
 
-size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+class SynchronizedCharQueue {
+  public:
+    vector<char> front() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock, [&]{ return !queue_.empty(); });
+        return queue_.front();
+    }
+
+    vector<char> pop() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock, [&]{ return !queue_.empty(); });
+        vector<char> item( queue_.front() );
+        queue_.pop();
+        return item;
+    }
+
+    void push(vector<char>&& items) {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        queue_.push(items);
+        mlock.unlock();
+        cond_.notify_one();
+    }
+
+    void push(const char& item) {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        queue_.push({item});
+        mlock.unlock();
+        cond_.notify_one();
+    }
+
+  private:
+    queue< vector<char> > queue_;
+    boost::mutex mutex_;
+    boost::condition_variable cond_;
+};
+
+struct UserData {
+    bool collecting_header;
+    vector<string> headers;
+    string content_type;
+    string error_response;
+    SynchronizedCharQueue stream;
+};
+
+size_t write_queue(char *ptr, size_t size, size_t nmemb, UserData& userdata) {
```

Comment:
Hmm, unless you're sure it's fine/portable, I'd suggest sticking to the curl spec here, i.e. "void *userdata" as the argument, and then coerce it internally, say `WriteQueueData *data = (WriteQueueData *) userdata;`, and as that suggests, I'd be inclined toward a name other than UserData.